nvme_gw_persistence.py

```diff
@@ -0,0 +1,443 @@
+            self.logger.error(f"Unable to write to omap: {ex}. Exiting!")
+            raise
+
+    def _delete_key(self, key: str):
+        """Deletes key from omap persistent config."""
+
+        version_update = int(self.version) + 1
+        with rados.WriteOpCtx() as write_op:
+            # Compare operation failure will cause delete failure
+            write_op.omap_cmp(self.OMAP_VERSION_KEY, str(self.version), 1)
+            self.ioctx.delete_omap_keys(write_op, (key,))
+            self.ioctx.set_omap(write_op, (self.OMAP_VERSION_KEY,),
+                                (str(version_update),))
+            self.ioctx.operate_write_op(write_op, self.omap_name)
+        self.version = version_update
+        self.logger.info(f"omap_key deleted: {key}")
+
+    def add_bdev(self, bdev_name: str, val: str):
+        """Adds a bdev to the persistent config."""
+        key = self.BDEV_PREFIX + bdev_name
+        self._write_key(key, val)
+
+    def delete_bdev(self, bdev_name: str):
+        """Deletes a bdev from the persistent config."""
+        key = self.BDEV_PREFIX + bdev_name
+        self._delete_key(key)
+
+    def _restore_bdevs(self, omap_dict):
+        """Restores a bdev from the persistent config."""
+
+        for (key, val) in omap_dict.items():
+            if key.startswith(self.BDEV_PREFIX):
+                args = self._clean_args(str(val, 'utf-8'))
+                self.logger.info(f"Restoring bdev: {args}")
+                try:
+                    return_string = self.spdk_rpc.bdev.bdev_rbd_create(
+                        self.spdk_rpc_client, args["ceph_pool_name"],
+                        args["rbd_name"], int(args["block_size"]))
+                    self.logger.info(f"Created bdev: {return_string}")
+                except Exception as ex:
+                    self.logger.error(f"bdev_rbd_create failed with: \n {ex}")
+                    raise
+
+    def add_namespace(self, subsystem_nqn: str, bdev_name: str, val: str):
+        """Adds a namespace to the persistent config."""
+        key = self.NAMESPACE_PREFIX + subsystem_nqn + "_" + bdev_name
+        self._write_key(key, val)
+
+    def delete_namespace(self, subsystem_nqn: str, bdev_name: str):
+        """Deletes a namespace from the persistent config."""
+        key = self.NAMESPACE_PREFIX + subsystem_nqn + "_" + bdev_name
+        self._delete_key(key)
+
+    def _restore_namespaces(self, omap_dict):
+        """Restores a namespace from the persistent config."""
+
+        for (key, val) in omap_dict.items():
+            if key.startswith(self.NAMESPACE_PREFIX):
+                args = self._clean_args(str(val, 'utf-8'))
+                self.logger.info(f"Restoring namespace: {args}")
+                try:
+                    return_string = self.spdk_rpc.nvmf.nvmf_subsystem_add_ns(
+                        self.spdk_rpc_client,
+                        args["subsystem_nqn"],
+                        args["bdev_name"],
```

Comment:
I don't think it will work in general case. `bdev_name` is what is returned by the spdk on `bdev_rbd_create` (e.g. "Ceph0"). How can you be sure that `bdev_rbd_create` that is called on restore will return the same `bdev_name` that was returned when it was called first time.

I have not looked in the spdk internals, but what about the case like this

bdev_rbd_create rbd/test0 # returns Ceph0
nvmf_subsystem_add_ns nqn0 Ceph0
bdev_rbd_create rbd/test1 # returns Ceph1
nvmf_subsystem_add_ns nqn1 Ceph1
nvmf_subsystem_delete nqn0
bdev_rbd_delete Ceph0

Now if we restart the gateway and it will try to restore, we get:

bdev_rbd_create rbd/test1 # returns Ceph0
nvmf_subsystem_add_ns nqn1 Ceph1 # error

It looks like we need to have/store internal bdev mapping and use it on restore.

Reply:
You are correct about this scenario. I would like to suggest a simpler solution, where we require the `--bdev` name parameter in the cli instead of allowing it to be optional as it is now. Then this name can be stored in the persistent config alongside the other bdev creation arguments.

Reply:
I have no strong opinion about requiring `--bdev` name parameter. From technical POV it does not look like it differs much from making the gateway to generate a unique name (internal id) itself (because it would need to store it either way), and I don't know if it will be convenient for users if they need to specify bdev name and make sure it does not collide. Probably yes, probably no. May be other people will have an opinion. Either way will work for me.